pkg/vpc/cleanup.go

```diff
@@ -30,49 +33,50 @@ func findAvailableNetworkInterfaces(provider api.ClusterProvider, spec *api.Clus
 			},
 		},
 	}
-	output, err := provider.EC2().DescribeNetworkInterfaces(input)
-	if err != nil {
-		return nil, errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
-	}
+
 	securityGroupRE, err := regexp.Compile(fmtSecurityGroupNameRegexForCluster(spec.Metadata.Name))
 	if err != nil {
-		return nil, errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
+		return errors.Wrap(err, "failed to create security group regex")
 	}
-	networkInterfaces := []string{}
-	for _, eni := range output.NetworkInterfaces {
-		id := *eni.NetworkInterfaceId
-		for _, sg := range eni.Groups {
-			if securityGroupRE.MatchString(*sg.GroupName) {
-				logger.Debug("found %q, which belongs to our security group %q (%s)", id, *sg.GroupName, *sg.GroupId)
-				networkInterfaces = append(networkInterfaces, id)
-				break
-			} else {
+
+	var lastErr error
+
+	err = ec2API.DescribeNetworkInterfacesPagesWithContext(ctx, input, func(output *ec2.DescribeNetworkInterfacesOutput, lastPage bool) bool {
+		for _, eni := range output.NetworkInterfaces {
+			id := *eni.NetworkInterfaceId
+			for _, sg := range eni.Groups {
+				if securityGroupRE.MatchString(*sg.GroupName) {
+					logger.Debug("found %q, which belongs to our security group %q (%s)", id, *sg.GroupName, *sg.GroupId)
+					if err := fn(id); err != nil {
+						lastErr = err
+						return false
+					}
+					break
+				}
 				logger.Debug("found %q, but it belongs to security group %q (%s), which does not appear to be ours", id, *sg.GroupName, *sg.GroupId)
-				break
 			}
 		}
+		return !lastPage
+	})
+
+	if err != nil {
+		return errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
 	}
-	return networkInterfaces, nil
+
+	return lastErr
```

Comment:
Why are we interested in the last error? Shouldn't we simply exit early on the first error?

Reply:
We do exit on error from the callback function `fn` https://github.com/weaveworks/eksctl/blob/f539f63495f13d76f70d5bc9aa2852986605b9ff/pkg/vpc/cleanup.go#L50 but that error needs to be propagated to the call to `DescribeNetworkInterfacesPagesWithContext` because the signature of its callback function (`func (output *ec2.DescribeNetworkInterfacesOutput, lastPage bool) bool`) does not support an error return value.

Reply:
Ah, I see, sorry I missed that. Would you mind renaming the variable so that it is less misleading? (it's not really the last error)

Reply:
Yeah, I agree that it's not that clear. What do you suggest renaming it to? I've seen `fnErr` used in some cases.

Reply:
`fnErr` would be fine, I think :)